### DISCUSSION TITLE
Can the TOC be integrated in the sidebar contents?

### DISCUSSION THREAD

**Discussion by pglpm** [quarto-dev/quarto-cli#5859] [Q&A] [answered]
### Description

I am preparing a book project for a course. Quarto is awesome for this!

The default HTML layout shows the list of chapters in the sidebar (left) and a TOC "On this page" relative to the present page in the margin (right).

Is there an option to have the TOC integrated within the sidebar? I mean that the relevant chapter entry in the sidebar would expand, showing all TOC sections. Here is a sort of visual explanation:
![moving_toc](https://github.com/quarto-dev/quarto-cli/assets/9976691/a4f1665f-92ed-48f2-aa2f-466a52cafa2a)

I have managed to move the TOC to the sidebar with the option `toc-location: left` in the `_quarto.yml` file, [as explained in the Guide](https://quarto.org/docs/reference/formats/html.html), but it appears under the list of chapters, separated from it.

I searched for other questions on this and found [this one](https://github.com/quarto-dev/quarto-cli/discussions/5361) to be somewhat related, but it isn't the same.

Just to justify this question: I personally find it logically neater if the left sidebar is dedicated to navigation- and site-related items, and the right margin only shows additional information related to the main, central text, such as margin notes, small plots, references, and similar. I'm really grateful that Quarto allows for this look out of the box!

Thank you!

**Comment by dragonstyle** [comment, marked as answer]
Currently we don't have a way to include the TOC as additional sub items in the sidebar (the sidebar is currently reserved for top level elements (e.g. pages, chapters, parts). It would be a little complicated since the TOC is really only known for a page once it is rendered and the headings are identified, but it is something we could consider for a future release...

Is it your preference in terms of layout appearance or is there some other use case driving this?

**Reply by pglpm** [reply]
Thank you. I imagined there could be difficulties integrating local & global information. I just wanted to make sure I had not missed about this possibility in the Guide. No big problem for me, I'm using `toc-location: left` so I have it under the list of top-level elements and it works fine.

It was just a preference. I'm relying on the margin to give extra information (footnotes, references, warnings,etc.). The TOC would appear when no information was present, and I found that a little confusing: I'd prefer the reader to instinctively know that "left=where I am in my reading" and "right=extra info on what I'm reading".

The Quarto Team is working on many new cool features, and I look forward to them – this has bottom priority :)

**Reply by andrewpbray** [reply]
I'm wondering: since the `toc-location: left` options results in the (document-level) TOC below the (project-level) TOC in the left sidebar, another option would be to tuck the document TOC up under the document title in the project TOC. @pglpm Would this be an improvement over the current functionality?

This would be similar to your original request; the only difference is that the document headers would only be visible once you're on the document page itself, not from other pages on the website. Seems like a little homespun piece of JS could do the trick: moving the On This Page up under the document title in the sidebar on page load. Perhaps the Quarto devs would consider this as a feature request though (`toc-location: project-contents`?) since it doesn't require the onerous processing of docs before compiling the project sidebar.

I find I also don't much like the `toc: sidebar-right` default. It makes my reading of the body feel cramped; I prefer all of the navigation in one place.

**Reply by pglpm** [reply]
Yes, that sounds like a good idea!

**Reply by jtbayly** [reply]
This is a decent idea for a workaround, but one of the problems with the TOC is that it is *entirely* unavailable on mobile. 

https://github.com/quarto-dev/quarto-cli/issues/5961

So you could javascript it up into the right spot, but then you also need to CSS it to visible.

However, the bigger problem is that it still only allows you to see the TOC of a book by visiting *every single chapter* of the book.

**Comment by ShKlinkenberg** [comment]
I would also prefer such behaviours. Coming from bookdown, I was immediately searching how to get the navigation in the sidebar.

**Reply by MarieAugerMethe** [reply]
I would also prefer this behaviour. Having it inside the sidebar gives more space for the main text and provides an easy way to see the sections and it's content.

**Comment by jtbayly** [comment]
I’m really hoping for this functionality as well. I want the TOC on the left to be for the whole book, like it was with the gitbook theme in bookdown. To the end user this is a *book,* not a collection of pages. It shouldn’t require navigating to a chapter to see what the sections are. 

I realize I could manually create the book’s complete TOC in the sidebar, but it feels like this should just be an option in the Yaml.

**Reply by jtbayly** [reply]
Actually, after testing, adding a “contents” section under “sidebar" to the _quarto.yml seems to be ignored for projects of type “book.”

So in order to make the site more like a book in terms of finding the right content from the TOC, I have to change it to type “website” and manually change the settings to get it to act like a book, and then hardcode all of the chapters and chapter sections into the yaml, which is not ideal.

**Comment by jtbayly** [comment]
I just realized this was not a feature request, so upvoting and discussing here doesn’t accomplish anything. For those that are interested, I just made a feature request that you can upvote: https://github.com/quarto-dev/quarto-cli/discussions/13392
